File: 2001/2/8-2.py
#!/usr/bin/env python

class Point:
    def __init__(self, r = 0, c = 0, p = None):
        self.row = r
        self.col = c
        if p is not None:
            self.row = p.row
            self.col = p.col

    def __cmp__(self, p):
        if self.row == p.row and self.col == p.col:
            return 0
        if self.row < p.row or self.col < p.col:
            return -1
        else:
            return 1

    def __add__(self, p):
        return Point(self.row + p.row, self.col + p.col)
        
    def __str__(self):
        return "(%s, %s)" % (self.row, self.col)

def isValid(p, dim):
    return p.row > 0 and p.row <= dim and p.col > 0 and p.col <= dim

fin = file("prob2.in")
UP = Point(-1, 0)
DOWN = Point(1, 0)
LEFT = Point(0, -1)
RIGHT = Point(0, 1)
while 1:
    line = fin.readline()
    if line == "":
        break
    line = line.split()
    gridDim = int(line[0])
    line = fin.readline().split()
    start = Point(int(line[0]), int(line[1]))
    line = fin.readline().split()
    end = Point(int(line[0]), int(line[1]))
    cnt = 0
    complete = 0
    visited = []
    visit = []
    grid = []
    for i in xrange(gridDim):
        line = fin.readline().split()
        ln = []
        for i in xrange(gridDim):
            ln.append(int(line[i]))
        grid.append(ln)
    visit.append(start)
    #print "\n".join([" ".join([str(i) for i in ln]) for ln in grid])
    while not complete:
        num = len(visit)
        cnt += 1
        for i in xrange(num):
            here = visit.pop(0)
            visited.append(here)
            if here == end:
                complete = 1
                break
            hereU = here + UP
            hereD = here + DOWN
            hereL = here + LEFT
            hereR = here + RIGHT
            if isValid(hereU, gridDim) and not visited.count(hereU):
                if grid[hereU.row - 1][hereU.col - 1] == 0:
                    visit.append(hereU)
            if isValid(hereD, gridDim) and not visited.count(hereD):
                if grid[hereD.row - 1][hereD.col - 1] == 0:
                    visit.append(hereD)
            if isValid(hereL, gridDim) and not visited.count(hereL):
                if grid[hereL.row - 1][hereL.col - 1] == 0:
                    visit.append(hereL)
            if isValid(hereR, gridDim) and not visited.count(hereR):
                if grid[hereR.row - 1][hereR.col - 1] == 0:
                    visit.append(hereR)
        if len(visit) == 0:
            break
    if complete:
        print "There are %s squares on a shortest path" % cnt,
    else:
        print "No paths were found",
    print "from %s to %s" % (start, end)

File: 2001/2/8-2.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <queue>
#include <set>
using namespace std;

struct Point
{
    int row, col;

    Point(int r = 0, int c = 0) : row(r), col(c)
    {}

    Point(const Point& p) : row(p.row), col(p.col)
    {}

    Point operator +(Point p)
    {
        return Point(row + p.row, col + p.col);
    }

    bool operator ==(Point p)
    {
        return ((row == p.row) && (col == p.col));
    }

    bool operator <(Point p)
    {
        return ((row < p.row) || (col < p.col));
    }

    friend ostream& operator <<(ostream& out, Point p)
    {
        out << "row: " << p.row << ", col: " << p.col;
        return out;
    }
};

bool operator <(const Point p, Point q)
{
    return ((p.row < q.row) || (p.col < q.col));
}

bool isValid(Point p, int dim)
{
    return ((p.row > 0) && (p.row <= dim) && (p.col > 0) && (p.col <= dim));
}

int main()
{
    ifstream in("prob2_in.txt");
    set<Point> visited;
    queue<Point> visit;
    int gridDim, num, cnt;
    Point start, end, here, hereU, hereD, hereL, hereR;
    const Point UP(-1, 0), DOWN(1, 0), LEFT(0, -1), RIGHT(0, 1);
    bool complete;
    while (in >> gridDim)
    {
        in >> start.row >> start.col >> end.row >> end.col;
        cnt = 0;
        complete = false;
        vector< vector<int> > grid(gridDim, vector<int>(gridDim, 0));
        visited.clear();
        while (!visit.empty())
            visit.pop();
        for (int i = 0; i < gridDim; i++)
            for (int j = 0; j < gridDim; j++)
                in >> grid[i][j];
        visit.push(start);
        while (!complete)
        {
            num = visit.size();
            cnt++;
            for (i = 0; i < num; i++)
            {
                here = visit.front();
                visited.insert(here);
                visit.pop();
                if (here == end)
                {
                    complete = true;
                    break;
                }
                hereU = here + UP;
                hereD = here + DOWN;
                hereL = here + LEFT;
                hereR = here + RIGHT;
                if (isValid(hereU, gridDim) && (visited.find(hereU) == visited.end()))
                {
                    if (grid[hereU.row - 1][hereU.col - 1] == 0)
                        visit.push(hereU);
                }
                if (isValid(hereD, gridDim) && (visited.find(hereD) == visited.end()))
                {
                    if (grid[hereD.row - 1][hereD.col - 1] == 0)
                        visit.push(hereD);
                }
                if (isValid(hereL, gridDim) && (visited.find(hereL) == visited.end()))
                {
                    if (grid[hereL.row - 1][hereL.col - 1] == 0)
                        visit.push(hereL);
                }
                if (isValid(hereR, gridDim) && (visited.find(hereR) == visited.end()))
                {
                    if (grid[hereR.row - 1][hereR.col - 1] == 0)
                        visit.push(hereR);
                }
            }
            if (visit.empty())
                break;
        }
        if (complete)
            cout << endl << "There are " << cnt << " squares on a shortest path";
        else
            cout << endl << "No paths were found";
        cout << " from (" << start.row << "," << start.col << ") to (" << end.row << ","
            << end.col << ")." << endl << endl;
    }
}
